alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/cognitiveagents/characteristics/cognitive/AbstractOrientingBehavior.kt

```diff
@@ -144,86 +159,93 @@ abstract class AbstractOrientingBehavior<T, P, A, N, E, M, F>(
     }
     private var state: State = State.START
 
-    override fun updateInternalStatus(curTime: Time?, executed: Boolean, env: Environment<T, *>?) = Unit
+    /**
+     * The position the pedestrian wants to reach, in absolute coordinates.
+     */
+    val desiredPosition: P get() = when (state) {
+        State.MOVING_TO_DOOR, State.CROSSING_DOOR, State.MOVING_TO_FINAL -> subdestination
+        else -> currentPosition
+    }
 
-    override fun getRate(): Double = timeDistribution.rate
+    private fun onStart(): Unit = when {
+        environment.graph().nodeContaining(currentPosition) != null -> state = State.NEW_ROOM
+        /*
+         * If the pedestrian cannot locate itself inside any room (unusual
+         * condition), it tries to reach the closest door/passage in order
+         * to enter one. If this isn't possible, it simply won't move.
+         */
+        else -> moveToClosestDoor(noDoorAction = { state = State.ARRIVED })
+    }
+
+    /*
+     * Picks the closest door among the provided ones (defaults to all doors)
+     * and sets all the state variables so as to move to the selected door.
+     * noDoorAction is called if no door could be found, defaults to nothing.
+     */
+    private fun moveToClosestDoor(
+        doors: Collection<F> = environment.graph().edgeSet(),
+        noDoorAction: () -> Unit = {}
+    ): Unit = doors
+        .map { it to crossingPoint(it) }
+        .minBy { it.second.distanceTo(currentPosition) }
+        ?.let { (closestDoor, crossingPoint) -> moveToDoor(closestDoor, crossingPoint) }
+        ?: noDoorAction.invoke()
```

Comment:
do you need invoke? Can't you just `noDoorAction()`?

Reply:
I can! Fixed.